[Coverage] Avoid emitting duplicate coverage mappings

vedantk opened the pull request:
- Only instrument closures and nested functions once.
- Only visit functions to build coverage mappings once (and assert that we don't create duplicate coverage mappings).

Addresses rdar://39200851 and rdar://39297172.

vedantk committed:
[Coverage] Only instrument ClosureExprs once

ClosureExprs should only be visited for profiling purposes once, when
the SILFunction definition for the closure is being emitted.

This fixes an issue where the coverage tooling can't figure out how to
attribute the code coverage data of a closure to the right function.

rdar://39200851

vedantk committed:
[Coverage] Assert that we don't emit duplicate coverage mappings

While generating SIL for a function with a default argument, we don't
need to emit two identical coverage mappings for the function body. The
same goes for functions which reference foreign functions.

This PR introduces an assertion which should catch similar problems in
the future.

rdar://39297172

vedantk commented:
@swift-ci Please smoke test

vedantk commented:
@swift-ci Please Test Source Compatibility

vedantk commented:
@swift-ci Please smoke test

vedantk commented:
@swift-ci Please smoke test

vedantk commented:
@swift-ci Please Test Source Compatibility

vedantk commented:
@swift-ci Please Test Source Compatibility

slavapestov commented on lib/SIL/SILProfiler.cpp:
```diff
@@ -185,8 +200,11 @@ struct MapRegionCounters : public ASTWalker {
   std::pair<bool, Expr *> walkToExprPre(Expr *E) override {
     if (auto *IE = dyn_cast<IfExpr>(E)) {
       CounterMap[IE->getThenExpr()] = NextCounter++;
-    } else if (isa<AutoClosureExpr>(E) || isa<ClosureExpr>(E)) {
```
After type checking, AutoClosureExpr and ClosureExpr are the same semantically, so whatever you do with one should be done with the other.

vedantk replied:
I'll try getting rid of the distinction between autoclosures and closures here and see if anything breaks.

vedantk replied:
Thanks for pointing this out, some nice simplifications fall out of not having special handling for autoclosures.

slavapestov commented on lib/SILGen/SILGen.cpp:
```diff
@@ -483,6 +483,10 @@ static bool hasSILBody(FuncDecl *fd) {
 }
 
+static bool haveProfiledAssociatedFuncDecl(SILDeclRef constant) {
+  return constant.isDefaultArgGenerator() || constant.isForeign;
```
What about stored property initializers?

vedantk replied:
I'm not sure. Isn't the decl associated with a stored property initializer some vardecl? If so, we wouldn't be at risk of instrumenting it twice.

slavapestov commented on lib/SIL/SILCoverageMap.cpp:
```diff
@@ -40,6 +41,13 @@ SILCoverageMap::create(SILModule &M, StringRef Filename, StringRef Name,
+                       }) &&
+         "Duplicate coverage mapping for function");
+
   M.coverageMaps.push_back(CM);
```
Why not use a DenseMap instead? The assertion is O(n)...

vedantk replied:
I thought the PR would be cleaner without this change. I'm planning on doing this as a follow-up, but can do it now if you'd prefer.

vedantk replied:
I'm going to look into the fallout from the source compatibility suite now. Let me know if you think I should make the DenseMap change in this PR (I suppose it'd have to be MapVector for deterministic printing).

vedantk committed:
[Coverage] Only instrument nested functions once

Coverage counters for a nested function can be assigned once (in its
parent's scope), and then again (in its own scope). Nested functions
should only be visited for coverage mapping purposes once.

This is related to r://39200851, which is the same bug but for closures.

vedantk committed:
[Coverage] Remove special handling of autoclosures

Treating AutoClosureExprs the same as ClosureExprs allows for some nice
code simplifications, and should be more robust.

vedantk committed:
[Coverage] Only instrument curried instance methods once

Another fix related to r://39297172, in which we avoid instrumenting the
function associated with a curried thunk more than once.

vedantk commented:
Ok, all of the source compatibility suite failures should be addressed.

vedantk commented:
@swift-ci Please smoke test

vedantk commented:
@swift-ci Please smoke test

vedantk commented:
@swift-ci Please Test Source Compatibility

vedantk commented:
@swift-ci Please Test Source Compatibility

vedantk committed:
Restore counter increments in closure bodies (squash me)

vedantk commented:
@swift-ci Please clean smoke test
